File: python/speed_tester.py
import datetime
import threading
import const
import baseline_test
import time


results = [None]*const.PassCount

class SpeedResult:
    def __init__(self, Duration, UnitSize, AllocationSpeed):
        self.Duration = Duration
        self.UnitSize = UnitSize
        self.AllocationSpeed = AllocationSpeed

class SpeedTester:
    def __init__(self, NoOutput, Duration, AllocationCount, UnitSize):
        self.NoOutput = NoOutput
        self.Duration = Duration
        self.AllocationCount = AllocationCount
        self.UnitSize = UnitSize

def NewSpeedTester():
    return SpeedTester(False, const.DefaultDuration, 0, const.DefaultUnitSize)

def NewWarmupSpeedTester():
    tester = NewSpeedTester()
    tester.NoOutput = True
    tester.Duration = 5 * const.DefaultDuration
    return tester

def _AllocatorThread(tester,index):
     global results
     results[index] = baseline_test.run_baseline(baseline_test.NewUnitAllocator(tester.Duration))

def run_speed_tester (tester):
    duration = tester.Duration.total_seconds()
    if not tester.NoOutput:
        print("Test settings: \n")
        print("Duration {duration} s \n".format(duration=duration))
        print("Thread Count {thread_count} \n".format(thread_count=threading.active_count()))
        print("Unit Size {unit_size} \n".format(unit_size=const.DefaultUnitSize))
    totalCount = 0
    
    
    for pass_num in range(const.PassCount):
        threads = [
                threading.Thread(target=_AllocatorThread, args=(tester, pass_num),daemon=True)
                for i in range(const.PassCount)
                ]
        for t in threads:
            t.start()
    
    tc = 0
    
    for allocator in results:
        tc += allocator.AllocationCount

    if totalCount < tc:
        totalCount = tc


    if not tester.NoOutput:
        print("Allocation speed \n")
        print("Operation per second {operation_num}".format(operation_num = totalCount/duration/10**6))
    
    return SpeedResult(duration, const.DefaultUnitSize, totalCount/duration/10**6)
    

File: Go/speed_test/speed_tester.go
package speed_test

import (
	. "../common"
	"fmt"
	"time"
)

const PassCount = 100

var DefaultDuration = time.Duration(time.Millisecond)
var DefaultUnitSize int64 = 16

type SpeedTester struct {
	NoOutput        bool
	Duration        time.Duration
	AllocationCount int64
	UnitSize        int64
}

func NewSpeedTester() *SpeedTester {
	t := &SpeedTester{
		NoOutput: false,
		Duration: DefaultDuration,
		UnitSize: DefaultUnitSize,
	}
	return t
}

func NewWarmupSpeedTester() *SpeedTester {
	t := NewSpeedTester()
	t.NoOutput = true
	t.Duration = time.Duration(5 * time.Millisecond)
	return t
}

func (t *SpeedTester) Run() float64 {
	duration := t.Duration.Seconds()
	if !t.NoOutput {
		fmt.Printf("Test settings:\n")
		fmt.Printf("  Duration:     %v ms\n", int64(t.Duration.Seconds()/Milli))
		fmt.Printf("  Thread count: %v\n", ThreadCount)
		fmt.Printf("  Unit size:    %v B\n", t.UnitSize)
		fmt.Println()
	}

	totalCount := int64(0)
	for pass := 0; pass < PassCount; pass++ {
		runner := NewParallelRunner(func(i int) IActivity { return NewUnitAllocator(t.Duration, t.UnitSize) })

		// runner.Run here will run the Run function in unit_allocator
		// in this function:

		// Within the given duration, allocate as StepSize * AllocationUnit as many times as possible

		activities := runner.Run()

		// Count the total number of allocationCount calculated by adding up the allocationcount in each thread
		allocators := make([]*UnitAllocator, 0)
		for _, a := range activities {
			allocators = append(allocators, a.(*UnitAllocator))
		}

		tc := int64(0)
		for _, a := range allocators {
			tc += a.AllocationCount
		}
		if totalCount < tc {
			totalCount = tc
		}
	}

	fmt.Printf("Allocation speed:\n")
	fmt.Printf("  Operations per second: %.3f M/s\n", float64(totalCount)/duration/Mega)
	//fmt.Printf("  Bytes per second:      %.3f GB/s\n", float64(totalCount * UnitSize)/duration/GB)
	fmt.Println()
	return float64(totalCount) / duration / Mega
}
